morpheus/controllers/rss_controller.py

```diff
@@ -33,123 +32,161 @@ class RSSController:
         fallback = False
+        cache_hit = False
 
-        if (feed["bozo"]):
-            cache_hit = False
-            fallback = True
+        if is_url:
+            response = self._session.get(url)
+            cache_hit = response.from_cache
+            feed_input = response.text
+        else:
+            feed_input = url
 
-            logger.info(f"Failed to parse feed: {url}. Trying to parse using feedparser directly.")
+        feed = feedparser.parse(feed_input)
 
-            # If that fails, use feedparser directly (cant cache this)
-            feed = feedparser.parse(url)
+        if feed["bozo"]:
+            cache_hit = False
 
-            if (feed["bozo"]):
-                raise RuntimeError(f"Invalid feed input: {url}. Error: {feed['bozo_exception']}")
+            if is_url:
+                fallback = True
+                try:
+                    logger.info(f"Failed to parse feed: {url}. Trying to parse using feedparser directly.")
+                    feed = feedparser.parse(url)
+                except Exception as ex:
+                    raise RuntimeError(f"Failed to parse feed using fallback: {url}: {ex}") from ex
+
+            if feed["bozo"]:
+                try:
+                    logger.warning("Failed to parse feed: %s, %s. Trying with other source", url,
+                                   feed['bozo_exception'])
+                    feed = self._try_parse_feed_with_beautiful_soup(url, is_url)
+                except Exception as exec_info:
+                    raise RuntimeError(f"Invalid feed input: {url}. Error: {exec_info}")
 
         logger.debug("Parsed feed: %s. Cache hit: %s. Fallback: %s", url, cache_hit, fallback)
 
         return feed
 
-    def parse_feeds(self):
+
+    def parse_feeds(self) -> feedparser.FeedParserDict:
         """
         Parse the RSS feed using the feedparser library.
 
         Returns
         -------
         feedparser.FeedParserDict
             The parsed feed content.
-
-        Raises
-        ------
-        RuntimeError
-            If the feed input is invalid or does not contain any entries.
         """
         for url in self._feed_input:
             try:
-                if (url in self._blacklisted_feeds):
+                if (url in self._errored_feeds):
                     continue
 
                 feed = self._try_parse_feed(url)
 
-                if (not feed.entries):
-                    raise RuntimeError(f"Invalid feed input: {url}. No entries found.")
-
                 yield feed
 
             except Exception as ex:
-                logger.warning(f"Failed to parse feed: {url}. The feed will be blacklisted and not retried.")
-
-                self._blacklisted_feeds.append(url)
+                logger.warning("Failed to parse feed: %s: %s. The feed will be not be retried.", url, ex)
+                self._errored_feeds.append(url)
 
-    def fetch_dataframes(self):
+    def fetch_dataframes(self) -> cudf.DataFrame:
```

Comment:
Dont annotate the type here. We are technically returning a `typing.Generator[cudf.DataFrame]`. The typing system will automatically figure out generators and coroutine return types for you

Reply:
Ah, you are correct. Will revert this. Thanks